### PR TITLE
p2p/simulations: use github.com/gorilla/websocket

### PR DESCRIPTION
**Pull request by fjl**
This package was the last remaining user of golang.org/x/net/websocket.
Migrating to the new library wasn't straightforward because it is no
longer possible to treat WebSocket connections as a net.Conn.

Changing the inproc adapter required access to rpc.newCodec. I've taken
the opportunity to clean up the codec API in package rpc so this function
could be exposed cleanly.

### CODE REVIEW

File: rpc/json.go [hunk 1]
```diff
@@ -164,43 +164,45 @@ func (c connWithRemoteAddr) RemoteAddr() string { return c.addr }
 // jsonCodec reads and writes JSON-RPC messages to the underlying connection. It also has
 // support for parsing arguments and serializing (result) objects.
 type jsonCodec struct {
-	remoteAddr string
+	remote string
```
Comment (karalabe) [inline rpc/json.go]: gofmt ;P

File: rpc/json.go [hunk 2]
```diff
@@ -164,43 +164,45 @@ func (c connWithRemoteAddr) RemoteAddr() string { return c.addr }
...
-	closed     chan interface{}          // closed on Close
+	closeCh     chan interface{}          // closed on Close
 	decode     func(v interface{}) error // decoder to allow multiple transports
 	encMu      sync.Mutex                // guards the encoder
 	encode     func(v interface{}) error // encoder to allow multiple transports
 	conn       deadlineCloser
 }
 
-func newCodec(conn deadlineCloser, encode, decode func(v interface{}) error) ServerCodec {
+// NewCodec creates a codec which uses the given functions to read and write. If conn
```
Comment (karalabe) [inline rpc/json.go]: `NewFuncCodec`

File: rpc/json.go [hunk 3]
```diff
@@ -164,43 +164,45 @@ func (c connWithRemoteAddr) RemoteAddr() string { return c.addr }
...
 	dec := json.NewDecoder(conn)
 	dec.UseNumber()
-	return newCodec(conn, enc.Encode, dec.Decode)
+	return NewFuncCodec(conn, enc.Encode, dec.Decode)
 }
 
-func (c *jsonCodec) RemoteAddr() string {
-	return c.remoteAddr
+func (c *jsonCodec) remoteAddr() string {
+	return c.remote
```
Comment (karalabe) [inline rpc/json.go]: if it's internal, we might as well just inline.

File: rpc/server.go [hunk 4]
```diff
@@ -66,13 +66,13 @@ func (s *Server) RegisterName(name string, receiver interface{}) error {
 	return s.services.registerName(name, receiver)
 }
 
-// ServeCodec reads incoming requests from codec, calls the appropriate callback and writes
+// receiver reads incoming requests from codec, calls the appropriate callback and writes
```
Comment (karalabe) [inline rpc/server.go]: `ServeCodec`?

File: rpc/json.go (hunk 3 above)
Comment (fjl) [reply rpc/json.go]: That's not possible because remoteAddr is an interface method.

**Pull request merged**